criu/tls.c

```
@@ -0,0 +1,343 @@
+# define GNUTLS_E_CERTIFICATE_VERIFICATION_ERROR GNUTLS_E_CERTIFICATE_ERROR
+#endif
+
+#undef	LOG_PREFIX
+#define LOG_PREFIX "tls: "
+
+#define CRIU_PKI_DIR SYSCONFDIR "/pki"
+#define CRIU_CACERT CRIU_PKI_DIR "/CA/cacert.pem"
+#define CRIU_CACRL CRIU_PKI_DIR "/CA/cacrl.pem"
+#define CRIU_CERT CRIU_PKI_DIR "/criu/cert.pem"
+#define CRIU_KEY CRIU_PKI_DIR "/criu/private/key.pem"
+
+#define SPLICE_BUF_SZ_MAX (PIPE_BUF * 100)
+
+#define tls_perror(msg, ret) pr_err("%s: %s\n", msg, gnutls_strerror(ret))
+
+static gnutls_session_t session;
+static gnutls_certificate_credentials_t x509_cred;
+static int tls_sk = -1;
+static int tls_sk_flags = 0;
+
+void tls_terminate_session()
+{
+	int ret;
+
+	if (!opts.tls)
+		return;
+
+	if (session) {
+		do {
+			/* don't wait for peer to close connection */
+			ret = gnutls_bye(session, GNUTLS_SHUT_WR);
+		}
+		while(ret == GNUTLS_E_AGAIN || ret == GNUTLS_E_INTERRUPTED);
+		gnutls_deinit(session);
+	}
+
+	tls_sk = -1;
+	if (x509_cred)
+		gnutls_certificate_free_credentials(x509_cred);
+}
+
+ssize_t tls_send(const void *buf, size_t len, int flags)
+{
+	int ret;
+
+	tls_sk_flags = flags;
+	do {
+		ret = gnutls_record_send(session, buf, len);
+	}
+	while (ret == GNUTLS_E_INTERRUPTED || ret == GNUTLS_E_AGAIN);
+	tls_sk_flags = 0;
+
+	if (gnutls_error_is_fatal(ret)) {
+		tls_perror("Failed to send data", ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+/*
+ * Read data from a file descriptor, then encrypt and send it with GnuTLS.
+ * This function is used for cases when we would otherwise use splice()
+ * to transfer data from PIPE to TCP socket.
+ */
+int tls_send_data_from_fd(int fd, unsigned long len)
+{
+	ssize_t copied;
+	unsigned long buf_size = min(len, (unsigned long)SPLICE_BUF_SZ_MAX);
+	void *buf = xmalloc(buf_size);
+
+	if (!buf)
+		return -1;
+
+	while (len > 0) {
+		copied = read(fd, buf, buf_size);
+		if (copied < 0) {
+			pr_perror("Can't read from pipe");
+			goto err;
+		}
+
+		if (copied == 0)
+			break;
```

need to report an error in this case